File: backend/app/nlp/predictor.py
import os
import re
import joblib
import pandas as pd

BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))
DATA_DIR = os.path.join(BASE_DIR, "data")

MODEL_PATH = os.path.join(DATA_DIR, "model.pkl")
EXT_PATH = os.path.join(DATA_DIR, "extractor.pkl")
ENC_PATH = os.path.join(DATA_DIR, "label_encoder.pkl")

model = joblib.load(MODEL_PATH)
extractor = joblib.load(EXT_PATH)
encoder = joblib.load(ENC_PATH)


# ================= CLEAN =================

def clean_text(text):
    text = text.lower()
    text = re.sub(r"[^a-z\s']", " ", text)
    return re.sub(r"\s+", " ", text).strip()


# ================= HAND FEATURES =================

NEG_KW = ['overwhelmed','exhausted','worthless','hopeless','empty','numb',
          'dark','cry','anxious','scared','fear','trapped','alone','lonely',
          'burden','guilt','shame','failure','useless','broken','pain']

POS_KW = ['happy','joy','grateful','peaceful','content','hope','love',
          'connected','supported','calm','strong','proud','better','good']


def hand_features(text):

    words = text.split()
    wc = max(len(words), 1)

    neg = sum(1 for w in words if any(k in w for k in NEG_KW))
    pos = sum(1 for w in words if any(k in w for k in POS_KW))

    return {
        "neg_kw": neg,
        "pos_kw": pos,
        "help_kw": 0,
        "negations": 0,
        "intensifiers": 0,
        "first_person": 0,
        "word_count": wc,
        "sent_ratio": (pos - neg) / wc,
        "neg_pos_diff": neg - pos
    }


# ================= CRISIS DETECTION =================

CRISIS_WORDS = [
    "kill myself",
    "want to die",
    "end my life",
    "hurt my baby",
    "harm my baby",
    "suicide"
]


def detect_crisis(text):

    t = text.lower()

    for word in CRISIS_WORDS:
        if word in t:
            return True

    return False


# ================= PREDICT =================

def predict(text):

    # Crisis detection first
    if detect_crisis(text):
        return {
            "severity": "Severe",
            "epds_score": 30,
            "alert": "CRISIS DETECTED"
        }

    c_text = clean_text(text)

    feats = hand_features(c_text)

    # EPDS estimate
    epds_score = max(
    0,
    min(
        30,
        int(feats["neg_kw"] * 2.5 - feats["pos_kw"] * 2)
    )
    )

    row = pd.DataFrame([
        {
            "clean": c_text,
            "epds_score": epds_score,
            **feats
        }
    ])

    X = extractor.transform(row)

    pred = model.predict(X)[0]
    label = encoder.inverse_transform([pred])[0]

    return {
        "severity": label,
        "epds_score": epds_score
    }

File: lib/logic/predictor.dart
import 'extractor.dart';

class PPDPredictor {
  final List<String> negKeywords = [
    'overwhelmed', 'exhausted', 'worthless', 'hopeless', 'empty', 'numb',
    'dark', 'cry', 'anxious', 'scared', 'fear', 'trapped', 'alone', 'lonely',
    'burden', 'guilt', 'shame', 'failure', 'useless', 'broken', 'pain'
  ];

  final List<String> posKeywords = [
    'happy', 'joy', 'grateful', 'peaceful', 'content', 'hope', 'love',
    'connected', 'supported', 'calm', 'strong', 'proud', 'better', 'good'
  ];

  final List<String> crisisWords = [
    "kill myself", "want to die", "end my life", 
    "hurt my baby", "harm my baby", "suicide"
  ];

  String cleanText(String text) {
    // Lowercase and remove non-alphabet characters
    String cleaned = text.toLowerCase().replaceAll(RegExp(r"[^a-z\s']"), " ");
    // Collapse extra whitespace
    return cleaned.replaceAll(RegExp(r"\s+"), " ").trim();
  }

  bool detectCrisis(String text) {
    String t = text.toLowerCase();
    return crisisWords.any((word) => t.contains(word));
  }

  HandFeatures calculateHandFeatures(String text) {
    List<String> words = text.split(" ");
    int wc = words.length > 0 ? words.length : 1;

    int negCount = 0;
    int posCount = 0;

    for (var word in words) {
      if (negKeywords.any((k) => word.contains(k))) negCount++;
      if (posKeywords.any((k) => word.contains(k))) posCount++;
    }

    return HandFeatures(
      negKw: negCount,
      posKw: posCount,
      wordCount: wc,
      sentRatio: (posCount - negCount) / wc,
      negPosDiff: negCount - posCount,
    );
  }

  Map<String, dynamic> predict(String text) {
    if (detectCrisis(text)) {
      return {
        "severity": "Severe",
        "epds_score": 30,
        "alert": "CRISIS DETECTED"
      };
    }

    String cleaned = cleanText(text);
    HandFeatures feats = calculateHandFeatures(cleaned);

    // Replicating the EPDS estimate logic
    int epdsScore = ((feats.negKw * 2.5) - (feats.posKw * 2)).toInt();
    epdsScore = epdsScore.clamp(0, 30);

    // Rule-based severity mapping (from existing logic)
    String severity = "Minimal";
    if (epdsScore >= 20) {
      severity = "Severe";
    } else if (epdsScore >= 15) {
      severity = "High";
    } else if (epdsScore >= 10) {
      severity = "Moderate";
    } else if (epdsScore >= 5) {
      severity = "Mild";
    }

    return {
      "severity": severity,
      "epds_score": epdsScore,
      "features": feats.toList(),
    };
  }
}
